lib/elixir/lib/exception.ex

```diff
@@ -1631,3 +1631,90 @@ defmodule ErlangError do
   defp nth(3), do: "3rd"
   defp nth(n), do: "#{n}th"
 end
+
+defmodule Inspect.Error do
+  @moduledoc """
+  Raised when a struct cannot be inspected.
+  """
+  # QUESTION: I am not sure if is it considered backward incompatible
+  # to enforce keys that did not exist before
+  @enforce_keys [:exception, :stacktrace, :struct]
+  defexception exception: nil,
+               message: nil,
+               stacktrace: nil,
+               struct: nil
+
+  @impl true
+  def exception(arguments) when is_list(arguments) do
+    with {:ok, exception} <- Keyword.fetch!(arguments, :exception) |> validate(:exception),
+         {:ok, stacktrace} <- Keyword.fetch!(arguments, :stacktrace) |> validate(:stacktrace),
+         {:ok, struct} <- Keyword.fetch!(arguments, :struct) |> validate(:struct),
+         {:ok, message} <- Keyword.get(arguments, :message) |> validate(:message) do
+      message =
+        if message do
+          message
+        else
+          message(%Inspect.Error{
+            exception: exception,
+            stacktrace: stacktrace,
+            struct: struct
+          })
+        end
+
+      %Inspect.Error{
+        exception: exception,
+        message: message,
+        stacktrace: stacktrace,
+        struct: struct
+      }
+    else
+      {:error, key, term} -> raise(ArgumentError, "invalid #{key}, got: " <> inspect(term))
+    end
+  end
+
+  defp validate(term, :exception) when is_exception(term), do: {:ok, term}
+  defp validate(term, :stacktrace) when is_list(term), do: {:ok, term}
+  defp validate(term, :struct) when is_struct(term), do: {:ok, term}
+  defp validate(term, :message) when is_binary(term) or is_nil(term), do: {:ok, term}
+  defp validate(term, key), do: {:error, key, term}
+
+  @impl true
+  def message(%__MODULE__{message: message}) when is_binary(message) do
+    message
+  end
+
+  def message(%__MODULE__{exception: exception, stacktrace: stacktrace, struct: struct})
+      when is_exception(exception) and is_list(stacktrace) and is_struct(struct) do
+    "got #{inspect(exception.__struct__)} with message:\n\n" <>
+      "    \"\"\"\n" <>
+      "#{pad(String.trim_trailing(Exception.message(exception), "\n"), 4)}\n" <>
+      "    \"\"\"\n\n" <>
+      "  while inspecting:\n" <>
+      pad(format_struct(struct), 4) <>
+      "\n"
+  end
+
+  # Helpers
+  def pad(message, padding_length) do
+    padding = String.duplicate(" ", padding_length)
+
+    result =
+      message
+      |> String.replace("\\n", "\n")
+      |> String.split("\n")
+      |> Enum.map(&(padding <> &1))
+      |> Enum.join("\n")
+      |> String.trim("\n")
```

Comment:
If you trim the input then you likely won't need to trim here nor when calling `pad`.

Reply:
You do,
```iex
iex(4)> "Foo\n" |> String.split("\n") |> Enum.join("\n")       
"Foo\n"
```
and in fact, many of those message ends in a new line.

Reply:
@josevalim  it is really good optimization, but I cannot apply it because I had to cover a case for empty lines, to keep them empty.
If you want to have a look, maybe there is room for optimization.
https://github.com/elixir-lang/elixir/pull/11403/commits/712010d0d7a7db10eb71183127cfa0c17682d9c4#diff-0e2e75b45f813c505f2fa6643a3cddd54e5443a43d16014eac12bc498dfbd050R8-R16